### PR TITLE
fix(grammar): reject glued instruction keywords (FROMalpine) via scanner word boundary

### PR DESCRIPTION
**Pull request by tinovyatkin**
## Summary

Closes the keyword word-boundary gap identified in #47, which was merged with `FROMalpine` documented as unfixable due to tree-sitter regex constraints. That verdict only holds for regex tokens — the external scanner can enforce the boundary instead, using the same zero-width-token trick tree-sitter-bash uses for `_concat`.

- **grammar.js**: new hidden `$._keyword_terminator` external, and a `keyword($, pattern, name)` helper wrapping all 20 instruction keywords plus `AS` and HEALTHCHECK's inner `CMD`.
- **src/scanner.c**: the terminator is a zero-width token that succeeds only when the character after the keyword is whitespace, a line end, or EOF. It is checked before `scan_line_continuation` (which destructively skips inline whitespace within the same scan call) and suppressed during error recovery so a zero-width match cannot stall the parser.

## Behavior

| Input | Before | After |
|---|---|---|
| `FROMalpine`, `RUNecho hi`, `EXPOSE8080`, `USERroot`, `MAINTAINERfoo` | silently valid | ERROR (matches Docker "unknown instruction") |
| `FROM alpine ASbuilder` | valid, alias `builder` | ERROR on `ASbuilder` |
| `FROM\` ⏎ `alpine` | valid | ERROR — BuildKit joins continuations before tokenizing, yielding `FROMalpine` |
| `FROM\` ⏎ ` alpine`, `FROM<TAB>alpine`, CRLF, backtick escape directive | valid | still valid |

The line-continuation semantics fall out naturally: when the lookahead is the escape char the scanner falls through, the continuation is consumed as an extra, and the boundary is re-checked at the start of the continued line — exactly where BuildKit's joined text resumes.

## Compatibility

The terminator is hidden and zero-width, so no visible tree shapes change: `node-types.json` is untouched and queries/bindings are unaffected.

## Testing

- `tree-sitter test`: 200/200 passing (191 existing + 9 new cases in `test/corpus/keyword_boundary.txt`)
- All 9 `examples/` Dockerfiles still parse with zero errors

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved Dockerfile parsing so instruction keywords must end at a proper boundary, preventing glued forms like `FROMalpine` or `RUNecho` from being accepted.
  * Tightened handling around `AS` in multi-stage builds to better detect invalid alias boundaries.
  * Preserved valid cases with tabs, whitespace, and supported line continuations after instruction keywords.

* **Tests**
  * Added parser coverage for keyword-boundary and line-continuation edge cases.

### CODE REVIEW

File: src/scanner.c [hunk 1]
```diff
@@ -809,6 +810,24 @@ bool tree_sitter_containerfile_external_scanner_scan(void *payload, TSLexer *lex
...
+    // inline whitespace and would move the lexer off the boundary. When the
+    // lookahead is the escape char, we fall through instead of failing so a
+    // line continuation is consumed as an extra first and the boundary is
+    // re-checked at the start of the continued line; this matches BuildKit's
+    // join-then-tokenize semantics (`FROM\<nl> alpine` is valid, while
+    // `FROM\<nl>alpine` joins to FROMalpine and is not). Suppressed during
+    // error recovery (ERROR_SENTINEL valid): zero-width tokens there can
+    // prevent the parser from making progress.
+    if (valid_symbols[KEYWORD_TERMINATOR] && !valid_symbols[ERROR_SENTINEL] &&
+        (is_inline_space(lexer->lookahead) || is_line_end(lexer))) {
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply src/scanner.c]: Fixed in e11140e. `is_inline_space` now guards the `iswspace` call with a named `MAX_WINDOWS_WCHAR_CODEPOINT` bound, so non-BMP codepoints do not reach `iswspace` on Windows/MSVC.

[the remark this answers is not in the document]
Comment (tinovyatkin) [reply src/scanner.c]: Fixed in e11140e. The keyword terminator scanner now looks through whitespace at the start of a continued physical line and refuses to treat whitespace before a comment-only line as the keyword boundary. Added corpus coverage for `FROM\\` followed by an indented comment-only continuation line.

**Comment by tinovyatkin** [comment]
@codex review

**Comment by tinovyatkin** [comment]
CI is green on e11140e; @codex review
